lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -422,23 +428,36 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
     }
     if (Proto1->isVariadic() != Proto2->isVariadic())
       return false;
-    if (Proto1->getExceptionSpecType() != Proto2->getExceptionSpecType())
+    
+    if (Proto1->getTypeQuals() != Proto2->getTypeQuals())
       return false;
-    if (Proto1->getExceptionSpecType() == EST_Dynamic) {
-      if (Proto1->getNumExceptions() != Proto2->getNumExceptions())
+    
+    // Check exceptions: This information may get lost in canonical type.
+    auto Spec1 = OrigProto1->getExceptionSpecType();
+    auto Spec2 = OrigProto2->getExceptionSpecType();
+    
+    if (!isUnresolvedExceptionSpec(Spec1) &&
+        !isUnresolvedExceptionSpec(Spec2)) {
+      if ((Spec1 != EST_DynamicNone || Spec2 != EST_BasicNoexcept) &&
```

Comment:
Why do we need this complex check here? Isn't `Spec1 != Spec2` sufficient?

Reply:
Also, exception specifications only part of the type system after C++17 (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0012r1.html). Maybe we should not require the exception specifications to be the same unless C++17 or above is used?

Reply:
If we have `foo () noexcept` in one TU and `foo ()` in another TU, is that an ODR violation?
Or that violates ODR only in C++17? I think it is hard to answer.

Reply:
For now: The "structurally equivalent" check is not an exact ODR check (it should check function body too and other things). It is to determine if two declarations "look similar". (I can not find "C11 6.2.7/1" to check what "structurally compatible layout" means here, `Sema::hasStructuralCompatLayout` is the single non-ASTImporter use.)